internal/file/file.go

```diff
@@ -0,0 +1,101 @@
+package file
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/spf13/afero"
+)
+
+const hashExt = ".sha256"
+
+// Copy copies a file from the given bytes to destination.
+func Copy(destination string, fin []byte) error {
+	if _, err := os.Stat(destination); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("could not stat file: %w", err)
+	}
+
+	fout, err := os.Create(destination)
+	if err != nil {
+		return fmt.Errorf("could not create file: %w", err)
+	}
+	defer fout.Close()
+
+	r := bytes.NewReader(fin)
+
+	if _, err := io.Copy(fout, r); err != nil {
+		return fmt.Errorf("could copy file: %w", err)
+	}
+
+	return nil
+}
+
+// CheckSize checks if a file exists and has an exact size.
+func CheckSize(filename string, size int) (bool, error) {
+	stat, err := os.Stat(filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("could not stat file: %w", err)
+	}
+	return stat.Size() == int64(size), nil
+}
+
+var testableFS = afero.NewOsFs()
+
+// CheckHash checks if a file has the given SHA256 hash.
+// It supports reading the file's current hash from a static file saved next to it with the hashExt extension.
+func CheckHash(filename, hash string) (bool, error) {
+	if fh, err := afero.ReadFile(testableFS, filename+hashExt); err == nil {
+		return string(fh) == hash, nil
+	}
+	fh, err := sha256Sum(testableFS, filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return true, err
+	}
+	return fh == hash, nil
+}
+
+func SaveHash(filename, hash string) error {
```

Comment:
There's no need to maintain the hash, as the file containing the hash can be tampered with as well.

Reply:
see other comments https://github.com/platformsh/cli/pull/94#discussion_r1303408451